File: 02_paradigm_exploration/recursion_and_backtracking/n_queen.scala
object NQueensApp {
  def solve(n: Int): List[List[Int]] = {
    def isSafe(col: Int, queens: List[Int], row: Int): Boolean =
      queens.zipWithIndex.forall {
        case (c, r) => c != col && math.abs(c - col) != row - r
      }

    def placeQueens(row: Int): List[List[Int]] = {
      if (row == n) List(Nil)
      else
        for {
          queens <- placeQueens(row + 1)
          col <- 0 until n
          if isSafe(col, queens, row)
        } yield col :: queens
    }

    placeQueens(0).map(_.reverse)
  }

  def printBoard(solution: List[Int]): Unit = {
    val n = solution.length
    solution.foreach { col =>
      println((0 until n).map(i => if (i == col) "Q" else ".").mkString(" "))
    }
  }

  def main(args: Array[String]): Unit = {
    val n = 8
    val start = System.nanoTime()
    val solutions = solve(n)
    val end = System.nanoTime()

    println(s"Total solutions: ${solutions.length}")
    println("First solution:")
    printBoard(solutions.head)
    println(f"Execution time: ${(end - start) / 1e6}%.3f ms")
  }
}

File: 02_paradigm_exploration/recursion_and_backtracking/n_queen.go
package main

import (
	"fmt"
	"time"
)

var solutions [][]int

func isSafe(queens []int, row, col int) bool {
	for r := 0; r < row; r++ {
		c := queens[r]
		if c == col || abs(c-col) == abs(r-row) {
			return false
		}
	}
	return true
}

func solveNQueens(n int, row int, queens []int) {
	if row == n {
		solution := make([]int, n)
		copy(solution, queens)
		solutions = append(solutions, solution)
		return
	}
	for col := 0; col < n; col++ {
		if isSafe(queens, row, col) {
			queens[row] = col
			solveNQueens(n, row+1, queens)
		}
	}
}

func printBoard(queens []int) {
	n := len(queens)
	for _, q := range queens {
		for c := 0; c < n; c++ {
			if c == q {
				fmt.Print("Q ")
			} else {
				fmt.Print(". ")
			}
		}
		fmt.Println()
	}
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func main() {
	n := 8
	queens := make([]int, n)
	start := time.Now()
	solveNQueens(n, 0, queens)
	elapsed := time.Since(start)

	fmt.Printf("Total solutions: %d\n", len(solutions))
	fmt.Println("First solution:")
	printBoard(solutions[0])
	fmt.Printf("Execution time: %s\n", elapsed)
}
